core/internal/testutils/testutils.go

```diff
@@ -166,38 +166,34 @@ func NewWSServer(t *testing.T, chainID *big.Int, callback JSONRPCHandler) (ts *t
 	ts.wsconns = make([]*websocket.Conn, 0)
 	handler := ts.newWSHandler(chainID, callback)
 	ts.s = httptest.NewServer(handler)
+	t.Cleanup(ts.Close)
 	return
 }
 
 func (ts *testWSServer) Close() {
-	ts.mu.Lock()
-	defer ts.mu.Unlock()
-	if ts.wsconns == nil {
-		ts.t.Log("Test WS server already closed")
-		return
-	}
-	ts.s.CloseClientConnections()
-	ts.s.Close()
-	for _, ws := range ts.wsconns {
-		ws.Close()
+	if func() bool {
+		ts.mu.Lock()
+		defer ts.mu.Unlock()
+		if ts.wsconns == nil {
+			ts.t.Log("Test WS server already closed")
+			return false
+		}
+		ts.s.CloseClientConnections()
+		ts.s.Close()
+		for _, ws := range ts.wsconns {
+			ws.Close()
+		}
+		ts.wsconns = nil // nil indicates server closed
+		return true
+	}() {
+		ts.wg.Wait()
```

Comment:
We need to wait for the wsconn goroutines to return, since they are still trying to use the test logger:
> panic: Log in goroutine after ...